Added 'physics' prop to Stepper.

soupjake opened the pull request:
Added 'physics' prop which allows the developer to assign the Stepper's scroll view's 'ScrollPhysics' which can solve with unwanted scrolling behaviour if the parent is also a scroll view.

Fixes https://github.com/flutter/flutter/issues/19513

Commit:
Update stepper.dart

Added 'physics' prop which allows the developer to assign the Stepper's scroll view's 'ScrollPhysics' which can solve with unwanted scrolling behaviour if the parent is also a scroll view. Defaults to 'AlwaysScrollableScrollPhysics' if null.

dnfield commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -12,6 +12,7 @@ import 'icons.dart';
 import 'ink_well.dart';
 import 'material.dart';
 import 'material_localizations.dart';
+import 'scroll_physics.dart';
```
This line should be removed - I'm pretty sure what you want is already handled by the import on line 5.

SnakeyHips replied:
Yeah you're right. I'll remove it.

dnfield commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -594,6 +605,7 @@ class _StepperState extends State<Stepper> with TickerProviderStateMixin {
 
     return ListView(
       shrinkWrap: true,
+      physics: widget.physics ?? AlwaysScrollableScrollPhysics(),
```
Are we sure we want to default this ot AlwaysScrollableScrollPhysics?  That seems off to me but I'm not sure I'm piecing all of this together correctly.

If it is right, it should be `const`.

SnakeyHips replied:
As other comment. Removing default.

dnfield commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -151,6 +153,15 @@ class Stepper extends StatefulWidget {
+  /// For example, determines how the scroll view continues to animate after the
+  /// user stops dragging the scroll view. 
+  ///
+  /// Defaults to [AlwaysScrollableScrollPhysics] but consider changing to 
```
Why are we defaulting this to anything?  Why not just keep it null to retain existing behavior?

SnakeyHips replied:
At the time I thought it would have been the better option to do but yeah with further testing it's better to keep to null. My bad!

Commit:
Update stepper.dart

Removed unnecessary import and defaulting to 'AlwaysScrollableScrollPhysics' if physics prop is null.

dnfield commented:
Analysis is still unhappy because of whitespace at the end of a line - can you fix that up?

dnfield commented:
Can you add a test that fails without this change?

Commit:
Update stepper.dart

Ran through format to remove unwanted whitespace which Analysis isn't happy about.

SnakeyHips commented:
Ran it through the VSCode Format command so hope that solved the whitespace issue. 

I'm not sure how to create a unit test to demonstrate why I'm proposing this PR but let's say we have a ListView including a Stepper:

```
ListView(
        children: <Widget>[
          Text('long text'),
          Stepper(
              //stuff here
              ),
          Text('more long text'),
```
The ListView within the Stepper will mess with the parent ListView resulting in the user being unable to scroll down the page:

![](https://i.imgur.com/NRPkdMF.gif)

With adding a `physics` prop to the Stepper, the developer is easily able to prevent this from happening by using another kind of `ScrollPhysics` such as `ClampingScrollPhysis` which results in this:

![](https://i.imgur.com/069aydz.gif)

You could achieve this using a `NestedScrollView` and setting up a `ScrollController` etc but I think this simple change results in the same result with it being far easier to implement as the developer.

dnfield commented:
We don't really have a lot of success with dartfmt on this repo, unfortunately.  It does some strange things with linebreaks that don't fit with the style of the repo.  VSCode actually has a setting to trim whitespace for you: 

![image](https://user-images.githubusercontent.com/8620741/51125950-bba36a80-17d6-11e9-936a-6e28d712586b.png)

Can you revert the dartfmt for this and reapply your other changes?

As far as a test goes, you should be able to create a `testWidgets` that creates your problematic scenario, uses a `tester.tap` or `tester.fling` or somesuch to try to scroll it, and verify what's visible on the screen.  With `physics` set to null, you should be able to `expect` that the first list item is still visible - with it set to `ClampingScrollPhysics` you should expect to see new items visible.

Commit:
Update stepper.dart

Tried reformatting again.

Commit:
Update stepper.dart

Tried reformatting again.

Commit:
Update stepper.dart

Formatting again why do you hate me github editor.

Commit:
Update stepper.dart

Commit:
Update stepper.dart

if this format doesn't work i'll cry

Commit:
Update stepper_test.dart

Added Stepper scroll tests. One that fails to find Text after Stepper if physics left as null and one that succeeds if physics set to `ClampingScrollPhysics()`

SnakeyHips commented:
Okay got the format right (finally) and added tests. Let me know if these are acceptable and thanks for helping (you can really tell this is my first PR to a Google project).

Commit:
Update stepper_test.dart

Added const constructors

Commit:
Update stepper_test.dart

trying to get rid of whitespace again

Commit:
Update stepper_test.dart

Commit:
Update stepper_test.dart

why whitespace why

Commit:
Update stepper_test.dart

Commit:
Update stepper_test.dart

Commit:
Update stepper_test.dart

Commit:
Update stepper_test.dart

Swapped to `findsNothing` because I'm an idiot.

HansMuller commented:
Sorry this has taken so long to reach the finish line! 

I'm going to fix up the last formatting problems and a note to the `physics` property doc about using ClampingScrollPhysics when the stepper is used within another scrolling container. And then, when the build is green I'll land the changes.

Thanks again for your contribution!

HansMuller committed:
Update stepper_test.dart

HansMuller committed:
Update stepper.dart

tomosullivan8 commented:
Can we get the Stepper's physics to also be implemented for `StepperType.horizontal`? I've changed it manually for now but presume the `physics` needs to be updated for the `ListView` where `StepperType.horizontal`?

DrFrancky commented:
Temp solution is to add:
`physics: widget.physics,`
after `Expanded(
          child: ListView(`
near line 671
